### PR TITLE
[DDW-254] Address validation upgrade to use cardano-addresses CLI via IPC

### PR DESCRIPTION
**Pull request by rhyslbw**
This PR upgrades the address validation feature to support all networks and styles, using the [cardano-addresses](https://github.com/input-output-hk/cardano-addresses) CLI, executed in a child process, with an IPC channel brokering the interaction. The feature requires the binary to be available on the PATH, which will be configured in the dev environment via nix.
 
## Todos

- [x] Add the binary to the nix-shells, and production bundles
- [x] Improve the render-side logic, ensuring the _network tag_ for each network is valid, as some placeholders have been used

## Screenshots

- Screenshot

---

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/messages/GGKFXSKC6)
- [ ] Test

---

### Basics

- [x] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [x] PR has a good description that summarizes all changes
- [x] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [x] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [x] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [x] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [x] There are no *flow* errors or warnings (`yarn flow:test`)
- [x] There are no *lint* errors or warnings (`yarn lint`)
- [x] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [x] There are no missing translations (running `yarn manage:translations` produces no changes)
- [x] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [x] Japanese text changes are proofread and approved (Junko Oda)
- [x] UI changes look good in all themes (Alexander Rukin)
- [x] Storybook works and no stories are broken (`yarn storybook`)
- [x] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [x] Important parts of the code are properly commented and documented
- [x] Code is properly typed with flow
- [x] React components are split-up enough to avoid unnecessary re-renderings
- [x] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

**Comment by ManusMcCole** [comment]
Hi @DeeJayElly and @rhyslbw. I have encountered a blocker in testing this today. Testnet and selfnode both started show this error today. Following the instructions means installing newer builds not containing the changes for this PR
image

**Comment by ManusMcCole** [comment]
Shelley Testnet build has constant node issues preventing it from syncing. Can you guys incorporate the STN 1.1.0 changes in to these builds ? 
image

**Comment by ManusMcCole** [comment]
Hi Guys,I am having difficuilty sending funds on Mainnet. It comes up as "please enter a valid address" when transaction should be valid.Wont send to either of the two valid addresses 
MainnetTransError

**Comment by ManusMcCole** [comment]
Same error on Shelley Testnet. Transaction should be valid 
Shelleytestnetinvalidtranserror-gif

**Comment by ManusMcCole** [comment]
I forgot to show the copy address part in the shelley testnet video but it was copied to my clipboard from the previous trial 
image

**Comment by DeeJayElly** [comment]
> Same error on Shelley Testnet. Transaction should be valid
> Shelleytestnetinvalidtranserror-gif

This is only happening on Mainnet? What about testnet @ManusMcCole ?

**Comment by ManusMcCole** [comment]
> This is only happening on Mainnet? What about testnet @ManusMcCole ?

Hi Aleksander, It happens on Daedalus testnet also

**Comment by rhyslbw** [comment]
@ManusMcCole The errors were most likely due to the underlying program not being in the env. The commits cherry-picked in since your last comment resolve this, so testing can proceed.

**Comment by ManusMcCole** [comment]
Hi @DeeJayElly and @rhyslbw sorry to have to report this again. Something in the build process must still be blocking or overriding your changes. I am receiving the same error as yesterday on Mainnet, Shelley Testnet and Daedalus Testnet so far
TransError25thJune

**Comment by ManusMcCole** [comment]
> Not ready yet @ManusMcCole . We will ping you when it is

Ok my bad. I read your comment in the email saying ready for retest and it doesnt send another email when you edited your comment so I proceeded with the information in the email

**Comment by rhyslbw** [comment]
@ManusMcCole The feature is now ready for testing if you have builds available

**Comment by gnpf** [comment]
@DeeJayElly I'm still reproducing this on Win:

lMK35eli6Y

**Comment by ManusMcCole** [comment]
Hi @rhyslbw and @DeeJayElly. Important to note that although Gabriela can produce this on Windows. I have checked 4 networks so far on Mac and no issue. All transactions that should send work successfully and all non-valid are blocked

**Comment by rhyslbw** [comment]
Thanks for confirming @ManusMcCole
@gnpf I'll improve the response so we're more accurately reflecting the error, as it's almost certainly an issue with the windows executable that performs the validation. From there we can address the root cause

**Comment by gnpf** [comment]
@DeeJayElly @rhyslbw @KtorZ @disassembler Still not working on Win 😢 
KytODVDqBh

**Comment by rhyslbw** [comment]
cardano-addresses-error

@disassembler I updated the main process error handling to better understand the failure. As indicated by the screenshot, it appears to be an older version of the exe included. Testing directly with the exe included by the Daedalus Bridge works as expected.

**Comment by ManusMcCole** [comment]
Hi @DeeJayElly  and @rhyslbw. The logic appears to be working a lot better. It is allowing correct transactions on Windows and disallowing incorrect ones so far. However how the errors are presented is not working as expected. In the screenshot below I am attempting to send real ada from mainnet to a Shelley Testnet Address. The transaction is blocked but there is no error message just the next button is greyed out 
image

**Comment by ManusMcCole** [comment]
When attempting to send real ada to a selfnode address. The error is presented a little differently with the please enter a valid address error surrounding the Amount box 
image

**Comment by ManusMcCole** [comment]
Hi @rhyslbw and @DeeJayElly. On flight build I am getting the white screen error 
image

**Comment by ManusMcCole** [comment]
When I try to send a transaction from Shelley Testnet to a Mainnet Address on Windows I get the error below. The error from before but also a calculate fee error. GUI error message is now just an error occured. See video below

transerrorandcalcfee

**Comment by ManusMcCole** [comment]
Here is a more clear shot of the console messages 
image

**Comment by ManusMcCole** [comment]
Also experiencing the white screen error in selfnode. The restored wallet is not there when Daedalus is restarted 
selfnodeezgif com-video-to-gif

**Comment by DeeJayElly** [comment]
> Hi @DeeJayElly and @rhyslbw. The logic appears to be working a lot better. It is allowing correct transactions on Windows and disallowing incorrect ones so far. However how the errors are presented is not working as expected. In the screenshot below I am attempting to send real ada from mainnet to a Shelley Testnet Address. The transaction is blocked but there is no error message just the next button is greyed out
> image

This is update now, pls retest it @ManusMcCole

**Comment by DeeJayElly** [comment]
> Also experiencing the white screen error in selfnode. The restored wallet is not there when Daedalus is restarted
> selfnodeezgif com-video-to-gif

Fixed, pls retest @ManusMcCole

**Comment by gnpf** [comment]
This message is displayed again when inserting valid addresses on Win 😢 

rmtPG1Fq9Y

**Comment by ManusMcCole** [comment]
On Windows Flight build creating and restoring a wallet leads to Daedalus freezing 
ScreenFrozen

**Comment by rhyslbw** [comment]
@ManusMcCole @DeeJayElly @gnpf @nikolaglumac 
I've disabled Win support in https://github.com/input-output-hk/daedalus/pull/2041/commits/8aa627f4f015ccb006d3a956ab4e4f3f394bff25
There's a bug with using `cardano-addresses` via the Node.js sub-proc in this env, despite it running fine directly in a shell. I will link this issue when the bug report is logged, but for now I suggest descoping it from this PR

**Comment by nikolaglumac** [comment]
@rhyslbw Windows is the most important platform for us. I suggest we keep this PR open until we get it working on Windows too.
cc @darko-mijic

**Comment by rhyslbw** [comment]
@nikolaglumac I agree it's the most important platform, just thought you might not want to hold up the others when the outcome for Windows is the same until https://github.com/input-output-hk/cardano-addresses/issues/40#issuecomment-656042686 is resolved

**Comment by nikolaglumac** [comment]
> @nikolaglumac I agree it's the most important platform, just thought you might not want to hold up the others when the outcome for Windows is the same until [input-output-hk/cardano-addresses#40 (comment)](https://github.com/input-output-hk/cardano-addresses/issues/40#issuecomment-656042686) is resolved

I stand by my original statement.

**Comment by rhyslbw** [comment]
@nikolaglumac reverted

**Comment by rhyslbw** [comment]
I'll be soon testing the potential fix being prepared in [open PR](https://github.com/input-output-hk/cardano-addresses/pull/44), to pair with my last commit https://github.com/input-output-hk/daedalus/commit/9f25d6850bca8151ff9bb9de5bda2601c25276f7

**Comment by KtorZ** [comment]
There's hope:

Screenshot from 2020-07-10 21-42-59

We've setup a Windows CI environment and all tests now pass on Windows.

**Comment by nikolaglumac** [comment]
@rhyslbw is this now done?

**Comment by rhyslbw** [comment]
@nikolaglumac Not yet, the version of `cardano-addresses` isn't in a branch we can target. I'll request the wallet team stage a branch for us to test

**Comment by rhyslbw** [comment]
@nikolaglumac https://github.com/input-output-hk/cardano-addresses/issues/40#issuecomment-657557597

**Comment by rhyslbw** [comment]
This feature now works as expected in the nix dev env on linux.
@gnpf @ManusMcCole please test out the packaged versions on all platforms. @DeeJayElly I can still see an unrelated error being thrown in the dev console, and I can't see an error handler for [this condition](https://github.com/input-output-hk/daedalus/blob/35742e4ea36e2475c6c17c050e76e980e80ee023/source/main/ipc/introspect-address.js#L26)

@nikolaglumac I've pushed a testing branch with the version of `cardano-addresses` required, however https://github.com/input-output-hk/cardano-wallet/pull/1902 will be the one that gets merged.

**Comment by ManusMcCole** [comment]
Hi @rhyslbw and @DeeJayElly.On Windows In some instances the correct error message invalid address is shown. Other times it just shows the Next button greyed out with this error message in the console. In this example I am attempting to send read Ada from a mainnet address to a Byron testnet address
ezgif com-video-to-gif (1)

**Comment by rhyslbw** [comment]
Thanks @ManusMcCole. I've added the error handler mentioned above:

> I can't see an error handler for _this_ condition

Please paste the error from the console when you test this next build

**Comment by gnpf** [comment]
@rhyslbw @DeeJayElly @nikolaglumac @KtorZ [Here's a document](https://docs.google.com/spreadsheets/d/1gb1gKpkEyYLYuN82u4Yxm_4iVBn3XFVsCjJuAj5kaHU/edit?usp=sharing) we're using on the QA team to track the results for this PR.
For the latest build (13584) we're getting the validation error for the input even on Mac. We'll continue updating the results on this sheet.

**Comment by nikolaglumac** [comment]
@rhyslbw @DeeJayElly @ManusMcCole @gnpf let's please get this one done today 🙏

**Comment by ManusMcCole** [comment]
Hi Guys, Not sure if this was ready for retest yet but still not working on Windows. Trying to send a valid transaction on Shelley Testnet but it fails. 
image

**Comment by rhyslbw** [comment]
It was working in my dev env before @DeeJayElly committed https://github.com/input-output-hk/daedalus/pull/2041/commits/e71e3e6d336e71d2c81e86190191f54ba48fbe8a
I will remove this so we can at least validate the feaure works

**Comment by rhyslbw** [comment]
@ManusMcCole @gnpf @nikolaglumac 
Apologies, that was my mistake. I've fixed an invalid error assertion, and found the source of issue that was leaving the empty error message in the console.

**Comment by gnpf** [comment]
@rhyslbw @DeeJayElly We've been testing mostly on Mac and Win and so far, the only network that still has [this validation issue](https://drive.google.com/file/d/1NXUd1p8QjPxB71p3wK4w4x1VVZ-Uapa3/view?usp=sharing) is staging (in both platforms). We'll let you know of further issues and continue updating the [sheet](https://docs.google.com/spreadsheets/d/1gb1gKpkEyYLYuN82u4Yxm_4iVBn3XFVsCjJuAj5kaHU/edit?usp=sharing). The validation is working properly on the rest of the networks.

cc: @ManusMcCole @mioriohk

**Comment by nikolaglumac** [comment]
@rhyslbw @DeeJayElly can you check the last issues reported by QA team?

**Comment by DeeJayElly** [comment]
> @rhyslbw @DeeJayElly can you check the last issues reported by QA team?

i am rechecking all of these stuff, we still have issues on staging & ITN Selfnode

File: source/renderer/app/stores/WalletsStore.js [hunk 1]
```diff
@@ -869,21 +875,43 @@ export default class WalletsStore extends Store {
...
+      expectedNetworkTag = STAGING_MAGIC;
+      validAddressStyles = ['Byron', 'Icarus'];
+    } else if (isIncentivizedTestnet && !isShelleyTestnet) {
+      expectedNetworkTag = 0;
+      validAddressStyles = ['Jormungandr'];
+    } else if (isTestnet) {
+      expectedNetworkTag = TESTNET_MAGIC;
+      validAddressStyles = ['Byron', 'Icarus'];
+    } else if (isShelleyTestnet) {
+      expectedNetworkTag = 0;
```
Comment (rhyslbw) [inline source/renderer/app/stores/WalletsStore.js:896]: @DeeJayElly This should be what you removed in https://github.com/input-output-hk/daedalus/pull/2041/commits/316d605b9bd60f5a6d7c60123dffeb5bdc326c62

**Comment by ManusMcCole** [comment]
Hi @rhyslbw and @DeeJayElly. I have transaction errors with Shelley Testnet and Staging on the new builds on MacOs so far. Shelley Testnet was working on the build previous to this. More info in available in the spreadsheet. We will continue to test and add to the spreadsheet https://docs.google.com/spreadsheets/d/1gb1gKpkEyYLYuN82u4Yxm_4iVBn3XFVsCjJuAj5kaHU/edit#gid=0

**Comment by gnpf** [comment]
@rhyslbw @DeeJayElly Most versions of the build 13683 were tested and the [sheet](https://docs.google.com/spreadsheets/d/1gb1gKpkEyYLYuN82u4Yxm_4iVBn3XFVsCjJuAj5kaHU/edit#gid=0) is updated. The issues with ITN and staging persisted.

CC: @ManusMcCole @mioriohk

**Comment by rhyslbw** [comment]
@gnpf The transaction failures are out of scope, so I've marked them as `Passed`. (I suspect you need to send more ADA now)
The issue with Shelley testnet is the address appears to be invalid. Comments left in the spreadsheet. The address should start with `addr_test`.

So the remaining builds we need to resolve are Jormungandr networks. I will follow up with the wallet devs

**Comment by DeeJayElly** [comment]
I confirm that Jormungandr - Staging address is resolved. Tested it multiple times with multiple addresses and its working.

As for the Shelley testnet the addresses used for testing are wrong.

**Comment by nikolaglumac** [comment]
@gnpf @ManusMcCole @mioriohk please approve this PR once you confirm ITN and Byron Staging address validation is working.

**Comment by ManusMcCole** [comment]
Hi @rhyslbw and @DeeJayElly. The Windows Shelley Testnet build is not working. These are the mnemonics.( twist, never,volcano, west,access,donate,simple, fan, render ,smile,elevator, vendor, deer, dolphin sign) and (solar,unusual,april,innocent,volcano,hover,shallow, witness, maximum, moral, loop, assault,above, churn, pull) 
image

**Comment by mioriohk** [comment]
> Hi @rhyslbw and @DeeJayElly. The Windows Shelley Testnet build is not working. These are the mnemonics.( twist, never,volcano, west,access,donate,simple, fan, render ,smile,elevator, vendor, deer, dolphin sign) and (solar, unusual, april, innocent, volcano, hover, shallow, witness, maximum, moral, loop, assault, above, churn, pull)
> image
Hi @rhyslbw and @DeeJayElly , I could replicate @ManusMcCole 's finding on MacOs Shelley Testnet build too. Tested on build 13685

**Comment by ManusMcCole** [comment]
Hi @rhyslbw and @DeeJayElly. I am getting the please enter a valid address error on Windows ITN - Rewards. Staging walley almost synced testing that next
image

**Comment by ManusMcCole** [comment]
Hi Guys. I got this error message trying to send transactions on Staging 
image

**Comment by ManusMcCole** [comment]
Hi Guys, Staging address validation works ignore my previous comment. I forgot that the next step Confirm Transaction Screen is out of scope for this PR

**Comment by gnpf** [comment]
@rhyslbw @DeeJayElly @nikolaglumac We updated the [sheet](https://docs.google.com/spreadsheets/d/1gb1gKpkEyYLYuN82u4Yxm_4iVBn3XFVsCjJuAj5kaHU/edit#gid=0) with the results for the build **13692**. Here's a summary:

- We found that the issue with **Shelley Testnet** is fixed. But I'd liked to mention, the addresses do not start `addr_test`. [Here's a screenshot](https://drive.google.com/file/d/19RpijUnPSbtDknCb3Zu7QKuAgAKUflUY/view) of the addresses retrieved by a new wallet created on Shelley Testnet.

- At the moment, we couldn't complete testing on Shelley QA, ITN rewards, QA and Nightly as we can't get them to sync.

- In **ITN Selfnode**, we found that Mac and Win persist [the old issue ](https://drive.google.com/file/d/1Kp_f87htOwxYR0lcNFELgHMACHdLmALM/view?usp=sharing)of not accepting valid addresses, [but in Linux is fixed](https://drive.google.com/file/d/1DPe6gEIKbpW4mx4Lg2bQRHe-Zu-oxpSY/view?usp=sharing). 

- We didn't find any issues with the validation on the rest of the builds.

**Comment by rhyslbw** [comment]
Thanks @gnpf 
> We found that the issue with Shelley Testnet is fixed. But I'd liked to mention, the addresses do not start addr_test. Here's a screenshot of the addresses retrieved by a new wallet created on Shelley Testnet.
Yes, and for this reason the validation logic was being rejected. The config was adjusted to match reality, so it's now passing for you. When this change comes into effect, we will revert.

**Comment by ManusMcCole** [comment]
Hi @DeeJayElly  and @. Unfortunatlely we still have the "Please enter a valid address error" on the Itn Selfnode Windows build :(

**Comment by nikolaglumac** [comment]
@gnpf @ManusMcCole @rhyslbw @DeeJayElly can we please have this one merged today? 🙏 
Is the ITN selfnode the only issue here?

**Comment by ManusMcCole** [comment]
> Is the ITN selfnode the only issue here?

Hi Nikola, Yes ITN Selfnode on Windows and Mac are the only issues right now. The latest build failed before it produced itn selfnode for Mac so we cant confirm if it is still an issue with the newest builds

**Comment by ManusMcCole** [comment]
Hi @DeeJayElly  and @rhyslbw. Im having trouble syncing 15 word ITN Wallets on the newest builds on Windows so far. Here is the error. It seems like it was a success but the close button remains frozen for some time and the x button. After a few clicks and waiting a while it appears to be restored correctly 
image

**Comment by ManusMcCole** [comment]
Also the transactions to correct addresses for Shelley Testnet on Windows are being blocked by the "Please enter a valid address" error 
image

**Comment by ManusMcCole** [comment]
i just seen there was a new build created just now. I will test that as soon as it is complete

**Comment by nikolaglumac** [comment]
@ManusMcCole ignore ITN builds. Focus on Shelley Testnet, Mainnet and Testnet (Byron).

**Comment by nikolaglumac** [comment]
@rhyslbw @DeeJayElly valid address on macOS Shelley Testnet (MC4) instance are rejected as invalid. Here is an address example which should be VALID: 
```
addr1qyjl2042q86lzjv0yr2xpuntuth53z2j7ekt2uc36d73wpw9t702582wh0guh6whxh3ur06dyc5pe9a3k9kwrzvwuhksr7uv7x
```

**Comment by ManusMcCole** [comment]
Valid Transactions blocked on WIndows Shelley Testnet. "Please enter a valid address" message is shown. Wallet restoration error is there intermittently

**Comment by DeeJayElly** [comment]
> @rhyslbw @DeeJayElly valid address on macOS Shelley Testnet (MC4) instance are rejected as invalid. Here is an address example which should be VALID:
> 
> ```
> addr1qyjl2042q86lzjv0yr2xpuntuth53z2j7ekt2uc36d73wpw9t702582wh0guh6whxh3ur06dyc5pe9a3k9kwrzvwuhksr7uv7x
> ```

@ManusMcCole @nikolaglumac I have updated the Shelley testnet network magic and the addresses are valid now. Pls retest it @ManusMcCole .

**Comment by nikolaglumac** [comment]
@ManusMcCole please focus ONLY on testing address validation.

**Comment by ManusMcCole** [comment]
Quick update. Shelley testnet transactions are working. Im trying to test Mainnet Flight but the wallet restoration is not working so I cannot test transactions. I will keep trying. The rest of the builds I need for windows testing were not successful

**Comment by ManusMcCole** [comment]
On mac Byron Testnet I am not able to send Ada from a Byron legacy wallet to a new Shelley 24 word wallet i just created due to the "please enter a valid address" error message ? Either through a transaction or through the move to a shelley compatible wallet feature.  Is this expected ? @nikolaglumac

**Comment by ManusMcCole** [comment]
Hi @DeeJayElly  and @rhyslbw. Unfortunately Testnet is not working on Linux or Mac. 
Screenshot 2020-07-28 at 23 18 34

**Comment by ManusMcCole** [comment]
Hi @rhyslbw and @DeeJayElly. Testnet on Mac is not allowing valid transactions
Screenshot 2020-07-29 at 09 32 54

**Comment by ManusMcCole** [comment]
Address validation is still failing on Mac Byron Testnet. See screenshot below 
image

**Comment by ManusMcCole** [comment]
Address validation is still failing on Mac Byron Testnet 
image

**Comment by ManusMcCole** [comment]
Hi @rhys and @DeeJayElly. Testnet still is not working :( Here are 2 examples that are not accepted addr1qrsmpl0edy69t9wd2alzkddtcm7ku7tp65sfrxpl9pf2wscagtmu0d337c8ngumzg5raqymyevrxay8swn56te32v6asjm2jft and addr1qzr0d5wy5454fdcm332pfl8kmxl9f6ezhmzx2yjdta8cgfgagtmu0d337c8ngumzg5raqymyevrxay8swn56te32v6as934s4d. 
image

**Comment by ManusMcCole** [comment]
Hi @Rhys and @DeeJayElly. Testnet still is not working :( Here are 2 examples that are not accepted addr1qrsmpl0edy69t9wd2alzkddtcm7ku7tp65sfrxpl9pf2wscagtmu0d337c8ngumzg5raqymyevrxay8swn56te32v6asjm2jft and addr1qzr0d5wy5454fdcm332pfl8kmxl9f6ezhmzx2yjdta8cgfgagtmu0d337c8ngumzg5raqymyevrxay8swn56te32v6as934s4d.
image

**Pull request merged**